Decimal.distance(to:) was mis-implemented

Dante-Broggi opened the pull request:
<!-- What's in this pull request? -->
Fix `Decimal.distance(to:)`
Add tests
Strip end of line whitespace

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-6785](https://bugs.swift.org/browse/SR-6785).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Dante-Broggi committed:
Add Tests for Decimal `Strideable` correctness

Dante-Broggi committed:
Editor whitespace

Dante-Broggi committed:
Actual Fix

Dante-Broggi committed:
more Editor whitespace

Dante-Broggi committed:
test the actual axiom of `Strideable`

xwu commented on test/stdlib/TestDecimal.swift:
```diff
@@ -240,6 +240,7 @@ class TestDecimal : TestDecimalSuper {
                 //Strideable methods
                 expectEqual(Decimal(i).advanced(by: Decimal(j), Decimal(i+j), "\(Decimal(i).advanced(by: Decimal(j)) == \(i) + \(j)"
                 expectEqual(Decimal(i).distance(to: Decimal(j), Decimal(j-i), "\(Decimal(i).distance(to: Decimal(j)) == \(j) - \(i)")
+                expectEqual(Decimal(i).advanced(by: Decimal(i).distance(to: Decimal(j)), Decimal(j), "Decimal(i).advanced(by: Decimal(i).distance(to: Decimal(j)) == Decimal(j)")
```
Some string interpolation is missing in the last argument there.

xwu commented on stdlib/public/SDK/Foundation/Decimal.swift:
```diff
@@ -24,13 +24,13 @@ extension Decimal {
 
     public static let pi = Decimal(_exponent: -38, _length: 8, _isNegative: 0, _isCompact: 1, _reserved: 0, _mantissa: (0x6623, 0x7d57, 0x16e7, 0xad0d, 0xaf52, 0x4641, 0xdfa7, 0xec58))
 
-    public var exponent: Int { 
```
Please revert any changes in whitespace that aren't related to the bug being fixed.

xwu commented on test/stdlib/TestDecimal.swift:
```diff
@@ -237,6 +237,10 @@ class TestDecimal : TestDecimalSuper {
                 expectEqual(Decimal(i*j), Decimal(i) * Decimal(j), "\(Decimal(i*j)) == \(i) * \(j)")
                 expectEqual(Decimal(i+j), Decimal(i) + Decimal(j), "\(Decimal(i+j)) == \(i)+\(j)")
                 expectEqual(Decimal(i-j), Decimal(i) - Decimal(j), "\(Decimal(i-j)) == \(i)-\(j)")
+                //Strideable methods
```
Nit: please follow the prevailing comment style in this file by putting a space between `//` and `Strideable`.

xwu reviewed (commented):
Nice catch! Just some minor comments.

Dante-Broggi committed:
Revert "more Editor whitespace"

This reverts commit b2d9390fec3beea8df7d3cff5af541877ff38bf1.

Dante-Broggi committed:
Revert "Editor whitespace"

This reverts commit d96b7ce2f8c15fb64bf65b6f9278089bc1c88df6.

Dante-Broggi committed:
format

Dante-Broggi committed:
missing parens

should have checked code syntax in Xcode

xwu commented on test/stdlib/TestDecimal.swift:
```diff
@@ -237,6 +237,10 @@ class TestDecimal : TestDecimalSuper {
+                // Strideable methods
+                expectEqual(Decimal(i).advanced(by: Decimal(j)), Decimal(i+j), "\(Decimal(i).advanced(by: Decimal(j))) == \(i) + \(j)")
+                expectEqual(Decimal(i).distance(to: Decimal(j)), Decimal(j-i), "\(Decimal(i).distance(to: Decimal(j))) == \(j) - \(i)")
+                expectEqual(Decimal(i).advanced(by: Decimal(i)).distance(to: Decimal(j)), Decimal(j), "Decimal(i).advanced(by: Decimal(i).distance(to: Decimal(j)) == Decimal(j)")
```
Still missing a paren within your string :)
And if you're following the style of earlier lines, consider escaping appropriate portions of each side of the equation in the string with `\()`.

huonw replied:
This currently looks like it's computing `j - (i + i)`, but maybe it should be `(i + j) - i`:

```swift
Decimal(i).distance(to: Decimal(i).advanced(by: Decimal(j)))
```

Or `i + (j - i)`:

```swift
Decimal(i).advanced(by: Decimal(i).distance(to: Decimal(j)))
```

Dante-Broggi committed:
Fixing actual correctness

I wrote the test in a non-syntax highlighting editor, fixed up the syntax in Xcode when it was posted out it was wrong, but didn't check the new version's correctness either. 😯
Hopefully I finally got it correct this time.

CodaFi commented:
This is wrong in corelibs as well. Let me generate a radar and put it on Foundation so they take a look here.

CodaFi commented:
For reference rdar://57273727

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

tbkka commented:
Closing:  This code is no longer in the Swift project.
